### PR TITLE
feat(node)!: multiple transactions

### PR DESCRIPTION
**Pull request by shumkov**
## Issue being fixed or feature implemented

To implement same block execution in JS Drive and support multiple proposals for the same block height RS Drive should expose multiple transaction interface 

## What was done?

- Updated `startTransaction`, `commitTransaction`, `abortTransaction`, `rollbackTransaction` to work with multiple transactions
- Updated all other methods to accept a transaction instance instead of `useTransaction` flag

## How Has This Been Tested?

With tests

## Breaking Changes

Node JS binding interface has changed

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

File: node/src/lib.rs [hunk 1]
```diff
@@ -8,38 +15,63 @@ use dash_abci::abci::messages::{
...
     "db is in readonly mode due to the active transaction. Please provide transaction or commit it";
 
-type DriveCallback = Box<dyn for<'a> FnOnce(&'a Platform, TransactionArg, &Channel) + Send>;
+type PlatformCallback =
+    Box<dyn for<'a> FnOnce(&'a Platform, &HashMap<usize, Transaction>, &Channel) + Send>;
 type UnitCallback = Box<dyn FnOnce(&Channel) + Send>;
+type TrasactionCallback =
+    Box<dyn FnOnce(Sender<PlatformWrapperMessage>, &Transaction, &Channel) + Send>;
+
+struct PlatformWrapperTransactionAddress(usize, Sender<PlatformWrapperMessage>);
```
Comment (shumkov) [inline node/src/lib.rs]: let's use fields

File: node/src/lib.rs [hunk 2]
```diff
@@ -8,38 +15,63 @@ use dash_abci::abci::messages::{
...
-    Callback(DriveCallback),
+    Callback(PlatformCallback),
     // Indicates that the thread should be stopped and connection closed
     Close(UnitCallback),
-    StartTransaction(UnitCallback),
-    CommitTransaction(UnitCallback),
-    RollbackTransaction(UnitCallback),
-    AbortTransaction(UnitCallback),
+    StartTransaction(TrasactionCallback),
+    CommitTransaction(usize, UnitCallback),
```
Comment (shumkov) [inline node/src/lib.rs]: Let's define type instead of usize?

File: node/src/lib.rs [hunk 3]
```diff
@@ -50,14 +82,16 @@ impl DriveWrapper {
...
-        let (tx, rx) = mpsc::channel::<DriveMessage>();
+        let (tx, rx) = mpsc::channel::<PlatformWrapperMessage>();
 
         // Create an `Channel` for calling back to JavaScript. It is more efficient
         // to create a single channel and re-use it for all database callbacks.
         // The JavaScript process will not exit as long as this channel has not been
         // dropped.
         let channel = cx.channel();
 
+        let tclone = tx.clone();
```
Comment (shumkov) [inline node/src/lib.rs]: Do we need to clone it twice?

File: node/src/lib.rs [hunk 4]
```diff
@@ -67,52 +101,84 @@ impl DriveWrapper {
...
-                    DriveMessage::CommitTransaction(callback) => {
-                        platform
-                            .drive
-                            .commit_transaction(transaction.take().unwrap())
-                            .unwrap();
+                    PlatformWrapperMessage::CommitTransaction(
+                        transaction_raw_pointer_address,
+                        callback,
+                    ) => {
+                        if let Some(transaction) =
```
Comment (shumkov) [inline node/src/lib.rs]: What will happen if `transaction_raw_pointer_address` is incorrect?

File: node/src/lib.rs [hunk 5]
```diff
@@ -226,22 +306,31 @@ impl DriveWrapper {
...
+            .send_to_drive_thread(move |platform: &Platform, transactions, channel| {
+                let transaction_address = maybe_boxed_transaction_address
+                    .expect("transaction address should be available");
+
                 platform
                     .drive
-                    .create_initial_state_structure(
-                        using_transaction.then(|| transaction).flatten(),
-                    )
+                    .create_initial_state_structure(transactions.get(&transaction_address))
```
Comment (shumkov) [inline node/src/lib.rs]: We need to throw an error in JS thread if transaction_address is incorrect.

File: node/src/lib.rs (hunk 3 above)
Comment (jawid-h) [reply node/src/lib.rs]: Yes

**Review by jawid-h — APPROVED** [review]

**Pull request merged**